plugin/mlsdbonlinelocator.cpp

```diff
@@ -0,0 +1,290 @@
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2.1 of the License.
+*/
+
+#include "mlsdbonlinelocator.h"
+
+#include <QtCore/QJsonDocument>
+#include <QtCore/QJsonObject>
+#include <QtCore/QJsonParseError>
+#include <QtCore/QVariantMap>
+#include <QtCore/QTextStream>
+#include <QtNetwork/QNetworkAccessManager>
+#include <QtNetwork/QNetworkRequest>
+#include <QtNetwork/QNetworkReply>
+#include <QtCore/QLoggingCategory>
+
+#include <sailfishkeyprovider.h>
+#include <qofonosimmanager.h>
+#include <qofonoextmodemmanager.h>
+#include <networkmanager.h>
+#include <networkservice.h>
+
+/*
+ * HTTP requests are sent based on the Mozilla Location Services API.
+ * See https://mozilla.github.io/ichnaea/api/geolocate.html for protocol documentation.
+ */
+
+Q_LOGGING_CATEGORY(lcGeoclueMlsdbOnline, "geoclue.provider.mlsdb.online")
+
+MlsdbOnlineLocator::MlsdbOnlineLocator(QObject *parent)
+    : QObject(parent)
+    , m_nam(new QNetworkAccessManager(this))
+    , m_modemManager(new QOfonoExtModemManager(this))
+    , m_simManager(0)
+    , m_networkManager(new NetworkManager(this))
+{
+    connect(m_nam, SIGNAL(finished(QNetworkReply*)), SLOT(requestOnlineLocationFinished(QNetworkReply*)));
+    connect(m_modemManager, SIGNAL(enabledModemsChanged(QStringList)), SLOT(enabledModemsChanged(QStringList)));
+    connect(m_modemManager, SIGNAL(defaultVoiceModemChanged(QString)), SLOT(defaultVoiceModemChanged(QString)));
+    connect(m_networkManager, SIGNAL(servicesChanged()), SLOT(networkServicesChanged()));
+}
+
+MlsdbOnlineLocator::~MlsdbOnlineLocator()
+{
+}
+
+void MlsdbOnlineLocator::networkServicesChanged()
+{
+    m_wifiServices = m_networkManager->getServices("wifi");
+}
+
+void MlsdbOnlineLocator::enabledModemsChanged(const QStringList &modems)
+{
+    Q_UNUSED(modems);
+    setupSimManager();
+}
+
+void MlsdbOnlineLocator::defaultVoiceModemChanged(const QString &modem)
+{
+    Q_UNUSED(modem);
+    setupSimManager();
+}
+
+bool MlsdbOnlineLocator::findLocation(const QList<MlsdbProvider::CellPositioningData> &cells)
+{
+    if (!loadMlsKey()) {
+        qCDebug(lcGeoclueMlsdbOnline) << "Unable to load MLS API key";
+        return false;
+    }
+
+    QNetworkRequest req(QUrl("https://location.services.mozilla.com/v1/geolocate?key=" + m_mlsKey));
+    req.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
+
+    QVariantMap map;
+    map.unite(globalFields());
+    map.unite(cellTowerFields(cells));
+    map.unite(wifiAccessPointFields());
+    map.unite(fallbackFields());
+
+    if (map.isEmpty()) {
+        // no field data available
+        qCDebug(lcGeoclueMlsdbOnline) << "No field data available for MLS online request";
+        return false;
+    }
+
+    QJsonDocument doc = QJsonDocument::fromVariant(map);
+    QByteArray json = doc.toJson();
+    QNetworkReply *reply = m_nam->post(req, json);
```

Comment:
You mean, have an m_currentRequest to track the request, and if non-null (i.e., old request is still outstanding) don't trigger a new request?  I agree, this sounds like a good idea.